Red - Rewrite

Twentysix26 opened the pull request:
A general and incomplete list of what changed and what is going to change with this ongoing rewrite.  
We are in the early stages, each of these points can be discussed, improved and even scrapped.  
Your feedback is very important :+1:

## Packages instead of cogs

This allows for multi file cogs, bringing a much better organization for large cogs.  
Each package can contain a variable amount of cogs and files. The user will now install packages.  
They will be contained in the "cogs" folder (to be renamed?).  
Each package will hold its own data inside its own folder.
The downloader will now install packages, not single cogs.

The juicy details on how the downloader will deal with this are still to be decided and up to discussion

## No more DataIO, aka how we deal with data

The old DataIO was nothing but a thin wrapper around some json functions. It was also not threadsafe, causing blocking issues.
We are now using Will's "config", #770.
This allows us to support a variety of backends (flat files, which will be default, and databases) and makes dealing with data much less of an hassle.

## Removed global admin and mod roles

They are pretty much unused. Every server sets their own roles when the bot gets invited, might as well remove them altogether.

## Removed email / password login support

~~I don't think it's worth supporting these when you can login with a token. A CLI flag is provided to allow user accounts to login.~~ I have been informed that the new discord.py doesn't support email / password at all, so this makes the decision even easier 👌 

## Co Owners

Yes, the common stance is that this is most of the time a pretty awful idea. And it is.  
It also has a legit use case: a hosted bot that multiple people have access to anyway.
I don't think it's fair not to implement this feature just to protect the clueless. There will of course be a lot of warnings.

## Re organization of standard packages

Still haven't thought much about it. I would want to avoid just a simple convertion of the old cogs and I'd like to bring something new to the table

## Both global and local blacklist / whitelist

This is already in and functioning, it just lacks the commands. I may end up extending this idea for complex permissions, like the widely used permissions cog offers

## Sharding

Lots of talk about sharding in recent times. We were ready to support DBs in this rewrite just to deal with the issue. At the end it turned out that discord.py can do [the heavy lifting](https://discordpy.readthedocs.io/en/rewrite/api.html#discord.AutoShardedClient) for us and we can stick to the old ways quite safely.

Twentysix26 committed:
First commit

Twentysix26 committed:
[Core] Moved checks.py to main core package

Twentysix26 committed:
[Core] Check for coowners in the db

Twentysix26 committed:
[Core] Handle connection errors, fetch token from env var / db

Also reimplemented the old status codes

Twentysix26 committed:
[Core] Save package status and restore it on boot

Also better error handling on package loading

Twentysix26 committed:
Added requirements.txt and LICENSE, removed duplicate file

Twentysix26 committed:
[Core] Properly end tasks and process flusher's queue on quit

palmtree5 commented:
I'd suggest adding in a prompt for when the bot joins a server to provide info to the server about the bot, including setting the mod/admin roles.

As for co-owners, perhaps have a "primary owner" who would be able to use things like [p]debug and make co-owners not able to use those sorts of things by default but have a toggle for it? Thinking here is that if someone wants to add a co-owner who doesn't have some sort of access to the computer the bot is running on, the co-owner wouldn't be able to use debug then but if someone was adding a co-owner who did have access, then having access to debug in that case would be fine

Kowlin commented on core/utils/helpers.py:
```diff
@@ -0,0 +1,200 @@
+      If True, any change to the "database" will be queued to the
+      flusher and scheduled for a later write
+    default_value: Optional=None
+      Same behaviour as a defaultdict
```
Possible rewording here, not everyone knows what a ``defaultdict`` was in its current context.

Twentysix26 replied:
Yeah it could be clearer

Kowlin commented:
Just going to (badly) quote Danny/Rapptz on this regarding the Json Flusher so we don't forget,

```
The most frequent cause of data corruption is when you run out of memory mid-write.
The operating system will just kill you at that point with no chance to recover safely.
If you're scheduling writes to be done in bulk (not a bad idea but this is actually not how databases do it, more on this later),
then if you die between the flushes you lost X frames of data instead of the last frame of data.
Databases use something called a transaction
It's also atomic.
You set up changes and it keeps track of it in a transation
Then when you're done with your changes you commit.
You can rollback a transaction too but that's something else
Transactions have to comply with ACID standards
So if you make a system you should ideally make it meet the same ones.
Anyway like I said earlier
A lock literally meets this lol
It's how it's usually implemented
Some RDBS allow control over the lock
Scheduling writes is a bad idea unless you don't care about the data.
So some data loss is acceptable
like spoops does with his presence data writes
he schedules it
and if it crashes it doesn't matter cause the data loss is fine
the presence data will re-do when he reboots(edited)
Also the rewrite automatically reconnects.
On every case.
```

Twentysix26 committed:
[Core] Removed flusher, helpers now save on changes

After some considerations, while the flusher was an interesting experiment, it doesn't seem to be reliable enough. It's been removed in favor of the helpers autosaving on any change in a threadsafe way.

Twentysix26 committed:
[Core] Implemented missing checks, extended DB

The DB has been extended and now covers the functionality of the old Settings class of the previous version

Twentysix26 committed:
[Core] Added relative_path parameter to json helpers

Since these helpers will mostly be used by packages, it's important to make sure that the data they create stays contained inside them. This also brings the additional benefit of being able to manipulate data inside a package without knowing the name of the package's folder itself

Twentysix26 commented:
If you already started experimenting with cogs watch out, the parameter `relative_path` that I just pushed defaults to True. Adjust your paths accordingly.

Twentysix26 committed:
[Core] Reworked intro and added oauth url

palmtree5 commented:
Would like to advocate to have something like `bot.session` which would be an instance of `aiohttp.ClientSession()` that all cogs could then use.

palmtree5 commented:
Also, a way of mass-saving data would be useful (for example, dumping json from a aiohttp request to a file via JsonDB)

palmtree5 commented:
Also (unless I'm missing something) there doesn't appear to be support for getting keys from nested dicts short of accessing the data directly

Kowlin commented:
A method to bulk add/remove keys into the JsonDB would be something I would like to see.

Twentysix26 committed:
[Core] Interactive configuration

Twentysix26 committed:
[Core] Added separate dev cog

This is enabled with the cli flag --dev

palmtree5 committed:
[General] Initial porting (#757)

Polls have been removed for now and they will be remade in the future (maybe even as a separate cog)

Chovin commented:
may not be even remotely the same in v3, but currently we have `[p]load filename` and `[p]help ClassName`
could we set a standard in v3 to let users refer to Cogs or Packages in the same way throughout the bot?

palmtree5 commented:
As for core cog conversion (for anyone trying to help with this):

- [x] Alias - Done in tekulvw/mongoV3_alias
- [ ] Audio - ~Probably~ Definitely being rewritten ~?~ (Will's working on it, but more than willing to hand it off)
- [ ] Customcom - 26 mentioned a complete rewrite ( #769 )
- [x] Downloader - Gonna rewrite it with cogs.red in mind (Will)
- [ ] Economy - #781 - Palm__
- [x] General - Converted ( #757 )
- [ ] Image - Rewritten ( #768 )
- [ ] Mod - In progress - Palm__
- [x] Owner - in progress? `core_commands` and `dev_commands` have some of the commands from v2 owner
- [ ] Streams - 26 mentioned a complete rewrite ( #769 )
- [ ] Trivia - Possible rewrite ( #769 )

Someone with the ability to edit please edit this as things are done or if I'm off base and got something wrong in this

tekulvw commented:
Do we want to bring Squid Permissions into the main bot? Or the concept at least, because if so I'll get to work on rewriting that as well.

tekulvw committed:
[Core] Replaced JsonDB with Config (#770)

Twentysix26 committed:
Added .tmp and .data to .gitignore

tekulvw committed:
Sentry error logging (#784)

Twentysix26 committed:
[Utils] Fixed pagify

Twentysix26 committed:
[Core] Ported [p]set commands, added [p]set coowners

Twentysix26 committed:
[Core] Fixed owner request's author not showing up

tekulvw committed:
[Core] Check for correct discord.py version at boot (#799)

Fixes #798

tekulvw committed:
[Config] Fix get() functionality (#815)

tekulvw committed:
[Core] Add Sentry logger for core cogs data gathering only (#813)

tekulvw committed:
[Dev] Added [p]repl and [p]eval (#814)

tekulvw committed:
[Core] Fixed [p]set owner and [p]set avatar (#817)

Twentysix26 committed:
[Dev] Fixed [p]repl, sanitize [p]eval output

tekulvw committed:
Added Downloader cog (#786)

Twentysix26 committed:
Fill out license info

Twentysix26 committed:
[Downloader] Cleaned up commands' docstrings, del -> delete

tekulvw committed:
[Core] Add dynamic versioning based on git tag (#790)

tekulvw committed:
Added Alias cog (#780)

Twentysix26 committed:
[Dev] Cast repl output to str before sending

Kowlin committed:
Added Dev_commands to the Sentry logger (#826)

Kowlin committed:
Revert "Added Dev_commands to the Sentry logger (#826)" (#831)

This reverts commit b483e24e3365e757bcf6068c605597b155f80802.

tekulvw committed:
[Downloader] [p]findcog: Find a cog from a command (#855)

tekulvw committed:
Added CODEOWNERS file (#848)

Twentysix26 committed:
[Core] Fixed [p]set owner

tekulvw committed:
[GitHub] Contributing guidelines (#864)

Add `CONTRIBUTING.md` to the project.

Twentysix26 committed:
[Core] Co-owners cli flag only, implemented owner cli flag

After giving it further thought, I don't believe co-owners really need to be a command, as it is unlikely that co-owners will need to be changed often enough.
Furthermore, this gives a further layer of protection to people who like to gloss over warnings.

Twentysix26 committed:
[Dev] server -> guild in dev commands

Kowlin committed:
[Sentry] Cleared out websocket breadcrumbs (#870)

This adds additional protections to make sure no sensitive information gets caught in our Sentry error logger.

tekulvw committed:
[Config] Rewrite (#869)

JoshB84 commented:
Hello, 
I would like to suggest multiple choice for trivia questions.
What color is the Sun?
A) White
B) Yellow
C) Purple
D) Blue

tekulvw commented:
This will be reopened at a later date containing the same information.

CreatorVilius commented:
k
